pkg/splunk/enterprise/afwscheduler_test.go

```diff
@@ -1690,6 +1738,100 @@ func TestPodCopyWorkerHandler(t *testing.T) {
 	close(ppln.pplnPhases[enterpriseApi.PhaseInstall].msgChannel)
 }
 
+func TestRunLocalScopedPlaybook(t *testing.T) {
+
+	// Test for each phase can send the worker to down stream
+	cr := enterpriseApi.ClusterMaster{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "ClusterMaster",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "stack1",
+			Namespace: "test",
+		},
+		Spec: enterpriseApi.ClusterMasterSpec{
+			AppFrameworkConfig: enterpriseApi.AppFrameworkSpec{
+				AppsRepoPollInterval:      60,
+				MaxConcurrentAppDownloads: 5,
+
+				VolList: []enterpriseApi.VolumeSpec{
+					{
+						Name:      "test_volume",
+						Endpoint:  "https://s3-eu-west-2.amazonaws.com",
+						Path:      "testbucket-rs-london",
+						SecretRef: "s3-secret",
+						Provider:  "aws",
+					},
+				},
+				AppSources: []enterpriseApi.AppSourceSpec{
+					{
+						Name:     "appSrc1",
+						Location: "adminAppsRepo",
+						AppSourceDefaultSpec: enterpriseApi.AppSourceDefaultSpec{
+							VolName: "test_volume",
+							Scope:   "local",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	var replicas int32 = 1
+	sts := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "splunk-stack1",
+			Namespace: "test",
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: &replicas,
+		},
+	}
+
+	// Create client and add object
+	c := spltest.NewMockClient()
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "splunk-stack1-clustermaster-0",
+			Namespace: "test",
+			Labels: map[string]string{
+				"controller-revision-hash": "v0",
+			},
+		},
+	}
+
+	// Add object
+	c.AddObject(pod)
+
+	// Just make the lint conversion checks happy
+	var client splcommon.ControllerClient = getConvertedClient(c)
+
+	var waiter sync.WaitGroup
+	waiter.Add(1)
+	var localInstallCtxt *localScopeInstallContext = &localScopeInstallContext{
+		worker: &PipelineWorker{
+			appSrcName:    cr.Spec.AppFrameworkConfig.AppSources[0].Name,
+			targetPodName: "splunk-stack1-clustermaster-0",
+			sts:           sts,
+			cr:            &cr,
+			appDeployInfo: &enterpriseApi.AppDeploymentInfo{
+				AppName: "app1.tgz",
+				PhaseInfo: enterpriseApi.PhaseInfo{
+					Phase:      enterpriseApi.PhaseDownload,
+					Status:     enterpriseApi.AppPkgDownloadPending,
+					RetryCount: 2,
+				},
+			},
+			afwConfig: &cr.Spec.AppFrameworkConfig,
+			client:    &client,
+			waiter:    &waiter,
+		},
+	}
+
+	localInstallCtxt.runPlaybook()
```

Comment:
so what is the criteria for this test case to pass?

Reply:
Good question. Being it is go routine, when writing the API return value didn't make sense, hence just covers the case, where the channel has to free upon exit(this is something we did for other areas of testing). To have enough coverage, returning values may not be a bad idea, then, again we have some limitations with the PodExecs.

Reply:
Probably, we can add a return value, and inspect whatever possible with the PodExec

Reply:
In my PR I am returning an error value for `runPlaybook` method. Also, with the PodExec thing, I have implemented a mock client which can be used to call the mock code instead of the actual PodExec code, which will help in covering all the cases in UTs.